pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,366 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kubestellar/kubestellar/api/edge/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// placementResolution stores the selected objects and destinations for a single placement.
+// The mutex should be read locked before reading, and write locked before writing to any field.
+type placementResolution struct {
+	sync.RWMutex
+
+	// map key is GVK/namespace/name as outputted by util.key.GvkNamespacedNameKey()
+	objectIdentifierToKey map[string]*util.Key
+	destinations          sets.Set[string]
+
+	// workloadGeneration is a local counter that reflects that internal
+	// changes occurred to the objects referenced in the objectIdentifierToKey
+	// map. That means, this field is incremented whenever an already noted
+	// object is called for noting again.
+	workloadGeneration int64
+}
+
+// noteObject adds/deletes an object to/from the resolution.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
+func (resolution *placementResolution) noteObject(obj runtime.Object) (bool, error) {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	key, err := util.KeyForGroupVersionKindNamespaceName(obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get key for object: %w", err)
+	}
+
+	// formatted-key to use for mapping
+	formattedKey := key.GvkNamespacedNameKey()
+	_, exists := resolution.objectIdentifierToKey[formattedKey]
+
+	// avoid further processing for keys of objects being deleted that do not have a deleted object
+	if isBeingDeleted(obj) {
+		delete(resolution.objectIdentifierToKey, formattedKey)
+
+		return exists, nil
+	}
+
+	if exists {
+		// the object is already in the resolution but the object content changed
+		resolution.workloadGeneration++
+	}
+
+	// add object to map
+	resolution.objectIdentifierToKey[formattedKey] = &key
+	// Internal changes to noted objects are also changes to the resolution.
+	return true, nil
+}
+
+// removeObject deletes an object from the resolution if it exists.
+// The return bool indicates whether the placement resolution was changed.
+// This function is thread-safe.
+func (resolution *placementResolution) removeObject(obj runtime.Object) bool {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	key, err := util.KeyForGroupVersionKindNamespaceName(obj)
+	if err != nil {
+		return false // assume object was never added
+	}
+
+	// formatted-key to use for mapping
+	formattedKey := key.GvkNamespacedNameKey()
+	_, exists := resolution.objectIdentifierToKey[formattedKey]
+
+	delete(resolution.objectIdentifierToKey, formattedKey)
+	return exists
+}
+
+// setDestinations updates the destinations list in the resolution.
+// Nothing in this function mutates the given set, and it is thread-safe.
+func (resolution *placementResolution) setDestinations(destinations sets.Set[string]) {
+	resolution.Lock()
+	defer resolution.Unlock()
+
+	resolution.destinations = destinations.Clone()
```

Comment:
Making a copy here works.
FYI, this package could be more efficient if it were the callers responsibility to provide a `Set[string]` that never gets mutated henceforth.